internal/middleware/user.go

```diff
@@ -47,34 +45,14 @@ func (u *UserSetter) setUser(ctx *gin.Context) {
 		ctx.Set(userContextKey, user)
 		return
 	}
-	user, err := u.persistedUser(username)
+	user, err := u.users.PersistedUser(username)
 	if err != nil {
 		log.Err(err).Msg("Failed to get user")
 		ctx.AbortWithStatus(http.StatusInternalServerError)
 		return
 	}
-	ctx.Set(userContextKey, user)
-}
-
-// Get a user that's persisted in the database by their username. If
-// they do not exist then they will be created with the base role
-func (u *UserSetter) persistedUser(username types.Username) (types.User, error) {
-	user := types.User{}
-	result := u.db.Where("username = ?", username).
-		Attrs(types.User{
-			Username: username,
-			Model:    types.Model{CreatedAt: time.Now()},
-		}).
-		FirstOrCreate(&user)
-	if result.Error != nil {
-		return user, types.NewErrServerError(result.Error)
+	if hasRole, err := rbac.HasRole(user, rbac.Base); err != nil && !hasRole {
+		_, _ = rbac.AddRole(user, rbac.Base)
```

Comment:
what situations are we expecting this to proc in? (since in PersistedUser if the user is created they should then be given base role)

Reply:
in the case the user has already been added with the wrong function, then tries to log in 🤦🏼